pkg/plugins/policies/meshhttproute/xds/configurer.go

```diff
@@ -21,27 +21,50 @@ type RoutesConfigurer struct {
 func (c RoutesConfigurer) Configure(virtualHost *envoy_route.VirtualHost) error {
 	matches := c.routeMatch(c.Matches)
 
-	var envoyRoutes []*envoy_route.Route
 	for _, match := range matches {
-		envoyRoutes = append(envoyRoutes, &envoy_route.Route{
-			Match: match,
+		r := &envoy_route.Route{
+			Match: match.getRouteMatch(),
 			Action: &envoy_route.Route_Route{
 				Route: c.routeAction(c.Clusters),
 			},
 			TypedPerFilterConfig: map[string]*anypb.Any{},
-		})
-	}
+		}
 
-	for _, envoyRoute := range envoyRoutes {
-		applyRouteFilters(c.Filters, envoyRoute)
+		// We pass the information about whether this match was created from
+		// a prefix match along to the filters because it's no longer
+		// possible to know for sure with just an envoy_route.Route
+		_, withPrefixMatch := match.(withPrefixMatch)
+		for _, filter := range c.Filters {
+			routeFilter(filter, r, withPrefixMatch)
+		}
+
+		virtualHost.Routes = append(virtualHost.Routes, r)
 	}
 
-	virtualHost.Routes = append(virtualHost.Routes, envoyRoutes...)
 	return nil
 }
 
-func (c RoutesConfigurer) routeMatch(matches []api.Match) []*envoy_route.RouteMatch {
-	var allEnvoyMatches []*envoy_route.RouteMatch
+type match interface {
+	getRouteMatch() *envoy_route.RouteMatch
+}
+
+type withPrefixMatch struct {
+	*envoy_route.RouteMatch
+}
+
+func (m withPrefixMatch) getRouteMatch() *envoy_route.RouteMatch { return m.RouteMatch }
+
+type withoutPrefixMatch struct {
+	*envoy_route.RouteMatch
+}
+
+func (m withoutPrefixMatch) getRouteMatch() *envoy_route.RouteMatch { return m.RouteMatch }
```

Comment:
Maybe I'm missing something, but if all implementations are the same, wouldn't it be better to have something like
```
type match struct{
  routeMatch *envoy_route.RouteMatch
  prefixMatch bool
}
```
?

Reply:
yes, got too hung up on using a sum type which are very painful in go.